packages/flutter_android_lifecycle/android/src/main/java/io/flutter/embedding/engine/plugins/lifecycle/FlutterLifecycleAdapter.java

```diff
@@ -0,0 +1,54 @@
+// Copyright 2019 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+package io.flutter.embedding.engine.plugins.lifecycle;
+
+import androidx.annotation.NonNull;
+import androidx.annotation.Nullable;
+import androidx.lifecycle.Lifecycle;
+import io.flutter.Log;
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+public class FlutterLifecycleAdapter {
+  private static final String TAG = "FlutterLifecycleAdapter";
+
+  @Nullable private Lifecycle lifecycle = null;
+
+  public FlutterLifecycleAdapter(@NonNull Object reference) {
```

Comment:
@mklim @matthew-carroll 

Currently the usage of this API looks like:

```java
      FlutterLifecycleAdapter adapter =
          new FlutterLifecycleAdapter(flutterPluginBinding.getLifecycle());
      Lifecycle lifecycle = adapter.getLifecycle();
```

What do you think about making the `getLifecycle` method static instead, I see how usually we want to avoid static methods for testability, but in this case I don't see how this is particularly useful especially without us adding some `FlutterLifecycleAdapterFactory` method which will potentially allow tests to replace the concrete implementation of FlutterLifecycleAdapter.

Second question - what do yo think about making the adapter be the one that calls `getLifecycle`?

If we do both of these we may end up with the call site looking like:

```java
  Lifecycle lifecycle = FlutterLifecycleAdapter.getLifecycle(flutterPluginBinding);
```

What do you think?

Reply:
I avoid any static call that I don't absolutely need, but it's up to you. Seems like you can also achieve the second goal whether you go static or not.

Reply:
> I avoid any static call that I don't absolutely need, but it's up to you.
I'm looking for feedback 😄 do you see any issue with using static here?

> Seems like you can also achieve the second goal whether you go static or not.
Yes, that's why I separated them to discuss separately, what is your opinion on that?

Reply:
I think the second suggestion is fine.

My feedback on the static suggestion is that I never make something static if I don't have to. I would personally not make it static. But I'm also not going to block this PR if you choose to do so. I don't see anything that will break by making it static. Then again, I could say the same thing for many instance methods that aren't static.

Reply:
This is totally subjective so I think it's up to you.

My personal opinion on `static` is that it's preferable if the method is truly immutable, doesn't require any kind of state or precondition, and _truly_ doesn't have any kind of side effects. Using `static` in that way is a soft indicator that a method is purely functional. Good examples are `Arrays#asList` or `TextUtils#isEmpty`. A "bad" counter example of `static` is our own FlutterJNI, which has a bunch of "static" methods that don't depend on any member variables but do depend on there being this magic global c library loaded in the program's global runtime. The danger in methods like that is first that it's extremely easy to call them out of sync, and then second that they're really hard to work around when testing.

The generic problem with any static method is that it can't be mocked in unit tests. I generally prefer testing real behavior to using mocks whenever possible so that's not a _huge_ concern to me but it can be a gigantic headache in the moment when the method really does need to be avoided in a test context (again, like `FlutterJNI`). Static also prevents subclasses from overriding the method, but that's a non-issue for me because I generally prefer composition to inheritance.

I think in this specific case it would be fine to make this static, because it's basically just returning the object that's passed into it. In a test context if you really needed to mock anything you would be returning a mock from the input parameter. I again prefer static interfaces because they're a soft signal that the function doesn't have silent side effects so I'd be happy to use one here, but it's totally subjective.

Reply:
updated to static, I think there's an upside (less cluttered call site) and no downside.